### ISSUE TITLE
Possible race-condition during Bot Command Registration

### ISSUE THREAD

**Issue by shehi** [discord-php/DiscordPHP#780]
**Environment**
- PHP Version:
    - 8.0
- DiscordPHP Version:
    - 7.0.8

**Describe the bug**
Following piece of code:
```
$discord->guilds[$discordServer->id]->commands->freshen()->done(static function (GuildCommandRepository $commands) use ($discord, $discordServer) {
    /** @var \App\Services\DiscordBotService\Commands\CommandInterface $commandClass */
    foreach (DiscordBotService::COMMANDS['guild'] as $commandClass) {
        $commandDefinition = $commandClass::build($discord);

        $existingCommand = $commands->get('name', $commandClass::NAME);
        if ($existingCommand) {
            $discord->guilds[$discordServer->id]->commands->delete($existingCommand);
        }

        $discord->guilds[$discordServer->id]->commands->save(
            new DiscordCommand($discord, $commandDefinition->toArray())
        );
    }
});
```

results in following log output:
```
[2022-03-31T08:56:58.472966+00:00] Discord Bot Logger.DEBUG: REQ GET applications/914614468741328896/commands successful [] []
[2022-03-31T08:56:58.473216+00:00] Discord Bot Logger.DEBUG: http not checking {"waiting":2,"empty":true} []
[2022-03-31T08:56:58.819301+00:00] Discord Bot Logger.DEBUG: REQ POST applications/914614468741328896/guilds/229980402574557184/commands successful [] []
[2022-03-31T08:56:58.820137+00:00] Discord Bot Logger.DEBUG: http not checking {"waiting":1,"empty":true} []
[2022-03-31T08:56:58.826041+00:00] Discord Bot Logger.DEBUG: REQ DELETE applications/914614468741328896/guilds/229980402574557184/commands/959012783616311326 successful [] []
[2022-03-31T08:56:58.826455+00:00] Discord Bot Logger.DEBUG: http not checking {"waiting":1,"empty":true} []
[2022-03-31T08:56:59.100020+00:00] Discord Bot Logger.DEBUG: REQ POST applications/914614468741328896/guilds/229980402574557184/commands successful [] []
[2022-03-31T08:56:59.100327+00:00] Discord Bot Logger.DEBUG: http not checking {"waiting":1,"empty":true} []
[2022-03-31T08:56:59.103068+00:00] Discord Bot Logger.DEBUG: REQ DELETE applications/914614468741328896/guilds/229980402574557184/commands/959012784887177266 successful [] []
```

As you can see, REQ DELETE is listed **after** REQ POST which should be the other way around. As a result, commands aren't being registered.

**THIS HAPPENS RANDOMLY!** One has to restart the Bot and hope this won't happen.

Sample correct log output, when commands are created successfully, looks like this:

```
[2022-03-31T09:02:44.547462+00:00] Discord Bot Logger.DEBUG: REQ GET applications/914614468741328896/commands successful [] []
[2022-03-31T09:02:44.547708+00:00] Discord Bot Logger.DEBUG: http not checking {"waiting":2,"empty":true} []
[2022-03-31T09:02:44.785960+00:00] Discord Bot Logger.DEBUG: REQ POST applications/914614468741328896/guilds/229980402574557184/commands successful [] []
[2022-03-31T09:02:44.786124+00:00] Discord Bot Logger.DEBUG: http not checking {"waiting":1,"empty":true} []
[2022-03-31T09:02:44.851426+00:00] Discord Bot Logger.DEBUG: REQ DELETE applications/914614468741328896/guilds/229980402574557184/commands/959013427475513354 successful [] []
[2022-03-31T09:02:44.851753+00:00] Discord Bot Logger.DEBUG: http not checking {"waiting":1,"empty":true} []
[2022-03-31T09:02:45.139595+00:00] Discord Bot Logger.DEBUG: REQ POST applications/914614468741328896/guilds/229980402574557184/commands successful [] []
```

**Additional context**
Main application is a Laravel 9 app. Bot is started via a Console Command.

**Comment by SQKo** [comment]
This is actually nature to how discord-php/http handles queue, depends on your rate limit bucket which may vary with http method.

First we all need to know (not just in discordphp, but in all Discord API libraries) that commands are Discord's data, treat them like channels, some people out there are trying to send create command request every time their bot start, that is misleading.

Second, those requests with delete() or save() are not immediately send, they are promised, async. I guess you know that, so what you are doing is quite not right.
You can make use of [Promise::all()](https://github.com/reactphp/promise#all) or [Promise::map()](https://github.com/reactphp/promise#map) to handle the queue of the request (i.e. ensuring a command is deleted first before created back), consult with the ReactPHP promise documentation for that.

In true practice, when dealing with bulk of commands, Discord provides these endpoints:
https://discord.com/developers/docs/interactions/application-commands#bulk-overwrite-global-application-commands
https://discord.com/developers/docs/interactions/application-commands#bulk-overwrite-guild-application-commands
The library does not utilizes those endpoints yet, but you can make them by using discord-php/http functions, e.g.
```php
$discord->getHttp()->put(Endpoint::bind(Endpoint::GUILD_APPLICATION_COMMANDS, $this->discord->application->id), $allyourcommands)->then(function ($response) {
    // do something with your commands data or update the repository
});
```
where `$allyourcommands` contains your command array structures or after converted the repository to array.
Basically this is like the opposite direction of `freshen()`, where we tell Discord to renew data of all the commands on their side from us.

**Comment by SQKo** [comment]
Also, we do think of making commands easier with requests auto generated, utilizing code reflection, portable Part class friendly to persistent storage and such, but right now application command from discord is not 100% complete, we might be expecting changes so we just currently follow their documented models.

**Comment by shehi** [comment]
:facepalm: Man, you are right. I missed the promise-based returns there due to the IDE not properly resolving the chain of methods in my code.

And thanks for the tip regarding HTTP functions we have. That approach is indeed more straight forward one.
